### PR TITLE
[Metal] Add bitmasked support in MetalRuntime

### PR DESCRIPTION
**Pull request by k-ye**
Related issue id = #593 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

As always, no rush on this. I'm simply sending out new ones as soon as the previous one gets merged. FYI I think we are pretty close on this :)

### CODE REVIEW

File: taichi/platform/metal/metal_runtime.cpp [hunk 1]
```diff
@@ -408,6 +492,83 @@ class MetalRuntime::Impl {
...
+    int list_data_mem_begin = 0;
+    for (int i = 0; i < max_snodes; ++i) {
+      auto iter = snode_descriptors.find(i);
+      if (iter == snode_descriptors.end()) {
+        continue;
+      }
+      const SNodeDescriptor &sn_meta = iter->second;
+      ListManager *rtm_list = reinterpret_cast<ListManager *>(addr) + i;
+      rtm_list->element_stride = sizeof(ListgenElement);
+      rtm_list->max_num_elems = sn_meta.total_num_elems_from_root;
```
Comment (yuanming-hu) [inline taichi/platform/metal/metal_runtime.cpp]: This should work fine for `bitmasked` but for sparser structures such as `ptr` and `hash`, `sn_meta.total_num_elems_from_root` might be really large. And then we need a dynamic list / memory allocator. Do you think that is doable on Metal? A basic allocator would only need
 - Preallocation of a huge chunk of GPU memory
 - A spinning lock, for e.g. the stack top of the allocator

Now I understand why starting from `bitmasked` is a good idea :-)

Comment (k-ye) [reply taichi/platform/metal/metal_runtime.cpp]: That's true. I didn't expect to handle memory allocation from Metal in the beginning, and hence the design...

We might have such an allocator with preallocated GPU memory + atomic int to store the head. I need some testings here. One thing I don't really know is how to best handle memory allocation failure on the device side... There's no crash or anything to convey that error to the host side, at least not without synchronization?

File: taichi/platform/metal/metal_runtime.cpp [hunk 2]
```diff
@@ -420,16 +581,18 @@ class MetalRuntime::Impl {
...
   ProfilerBase *const profiler_;
-  const bool needs_root_buffer_;
   nsobj_unique_ptr<MTLDevice> device_{nullptr};
   nsobj_unique_ptr<MTLCommandQueue> command_queue_{nullptr};
   nsobj_unique_ptr<MTLCommandBuffer> cur_command_buffer_{nullptr};
   std::unique_ptr<BufferMemoryView> root_mem_{nullptr};
   nsobj_unique_ptr<MTLBuffer> root_buffer_{nullptr};
   std::unique_ptr<BufferMemoryView> global_tmps_mem_{nullptr};
   nsobj_unique_ptr<MTLBuffer> global_tmps_buffer_{nullptr};
+  std::unique_ptr<BufferMemoryView> runtime_mem_{nullptr};
```
[the remark this answers is not in the document]
Comment (k-ye) [reply taichi/platform/metal/metal_runtime.cpp]: Ack, will work on it after work...

File: taichi/platform/metal/metal_runtime.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/platform/metal/metal_runtime.cpp]: > One thing I don't really know is how to best handle memory allocation failure on the device side... There's no crash or anything to convey that error to the host side, at least not without synchronization?

For CUDA without unified memory, I'll just call an `__assertfail` intrinsic function to abort the GPU threads. I guess there might be something similar on Metal. And you are right, as long as there's no synchronization, convey the error to host is difficult. I think at least we should inform the user of OOM in debug mode when a sync is done after every kernel.

File: taichi/platform/metal/metal_runtime.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (k-ye) [reply taichi/platform/metal/metal_runtime.cpp]: Done!

File: taichi/platform/metal/metal_runtime.cpp (hunk 1 above)
Comment (k-ye) [reply taichi/platform/metal/metal_runtime.cpp]: Added a comment..

**Pull request merged**